tests/unit/test_kernel.py

```diff
@@ -202,22 +208,375 @@ def expected_divergence_x_grad_y(
+
+        if dimension != 1:
+            expected_gradient = jnp.tile(expected_gradient, num_points**2).reshape(
+                num_points, num_points, self.mock_dimension
+            )
+
+        return expected_gradient
+
+    def expected_divergence_x_grad_y(
+        self, x: ArrayLike, y: ArrayLike, kernel: AdditiveKernel
+    ) -> np.ndarray:
+        num_points = x.shape[0]
+
+        expected_divergences = jnp.tile(
+            kernel.first_kernel.divergence_x_grad_y_elementwise(x, y)
+            + kernel.second_kernel.divergence_x_grad_y_elementwise(x, y),
+            num_points**2,
+        ).reshape(num_points, num_points)
+
+        return expected_divergences
+
+    def test_kernel_calls_sub_kernels_correctly(
+        self, kernel: _MultiCompositeKernel
+    ) -> None:
+        """Ensure AdditiveKernel calls sub kernel methods correctly."""
+        x = jnp.array(1.0)
+        y = jnp.array(1.0)
+
+        # Variable rename allows for nicer automatic formatting
+        k = self.reset_mocked_sub_kernels(kernel)
+
+        k = self.reset_mocked_sub_kernels(kernel)
+        k.compute_elementwise(x, y)
+        k.first_kernel.compute_elementwise.assert_called_once_with(x, y)
+        k.second_kernel.compute_elementwise.assert_called_once_with(x, y)
+
+        k = self.reset_mocked_sub_kernels(kernel)
+        k.grad_x_elementwise(x=x, y=x)
+        k.first_kernel.grad_x_elementwise.assert_called_once_with(x, y)
+        k.second_kernel.grad_x_elementwise.assert_called_once_with(x, y)
+
+        k = self.reset_mocked_sub_kernels(kernel)
+        k.grad_y_elementwise(x=x, y=x)
+        k.first_kernel.grad_y_elementwise.assert_called_once_with(x, y)
+        k.second_kernel.grad_y_elementwise.assert_called_once_with(x, y)
+
+        k = self.reset_mocked_sub_kernels(kernel)
+        k.divergence_x_grad_y_elementwise(x, y)
+        k.first_kernel.divergence_x_grad_y_elementwise.assert_called_once_with(x, y)
+        k.second_kernel.divergence_x_grad_y_elementwise.assert_called_once_with(x, y)
+
+
+class TestProductKernel(
+    BaseKernelTest[ProductKernel],
+    KernelMeanTest[ProductKernel],
+    KernelGradientTest[ProductKernel],
+    KernelCompositionTest[ProductKernel],
+):
+    """Test ``coreax.kernel.ProductKernel``."""
+
+    # Set up the size and dimension of the "dataset" that the vectorised
+    # mocked kernel methods act on
+    mock_num_points: int = 5
+    mock_dimension: int = 3
+
+    @override
+    @pytest.fixture(scope="class")
+    def kernel(self) -> _MultiCompositeKernel:
+        """Define an ProductKernel composed of mocked kernel functions."""
+        # Variable rename allows for nicer automatic formatting
+        num_points, dimension = self.mock_num_points, self.mock_dimension
+
+        k1 = MagicMock(spec=Kernel)
+        k1.compute_elementwise.return_value = jnp.array(1.0)
+        k1.compute.return_value = 1.0 * jnp.ones((num_points, num_points))
```

Comment:
Why do all these multiply `jnp.ones` by `1.0`?

Reply:
It was a way to cast to float without having the line be split over multiple lines but I've changed this to using jnp.full now so not an issue.